File: app/src/main/java/code/name/monkey/retromusic/loaders/SearchLoader.java
package code.name.monkey.retromusic.loaders;

import android.content.Context;
import android.support.annotation.NonNull;
import android.text.TextUtils;
import java.util.ArrayList;

import code.name.monkey.retromusic.R;
import io.reactivex.Observable;

public class SearchLoader {

    public static Observable<ArrayList<Object>> searchAll(@NonNull Context context, @NonNull String query) {
        ArrayList<Object> results = new ArrayList<>();
        return Observable.create(e -> {
            if (!TextUtils.isEmpty(query)) {
                SongLoader.getSongs(context, query)
                        .subscribe(songs -> {
                            if (!songs.isEmpty()) {
                                results.add(context.getResources().getString(R.string.songs));
                                results.addAll(songs);
                            }
                        });

                ArtistLoader.getArtists(context, query)
                        .subscribe(artists -> {
                            if (!artists.isEmpty()) {
                                results.add(context.getResources().getString(R.string.artists));
                                results.addAll(artists);
                            }
                        });
                AlbumLoader.getAlbums(context, query)
                        .subscribe(albums -> {
                            if (!albums.isEmpty()) {
                                results.add(context.getResources().getString(R.string.albums));
                                results.addAll(albums);
                            }
                        });
            }
            e.onNext(results);
            e.onComplete();
        });
    }
}

File: app/src/main/java/code/name/monkey/retromusic/loaders/SearchLoader.kt
/*
 * Copyright (c) 2019 Hemanth Savarala.
 *
 * Licensed under the GNU General Public License v3
 *
 * This is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by
 *  the Free Software Foundation either version 3 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 */

package code.name.monkey.retromusic.loaders

import android.content.Context
import code.name.monkey.retromusic.R
import code.name.monkey.retromusic.model.Genre
import java.util.*

object SearchLoader {
    fun searchAll(context: Context, query: String?): MutableList<Any> {
        val results = mutableListOf<Any>()
        query?.let { searchString ->
            val songs = SongLoader.getSongs(context, searchString)
            if (songs.isNotEmpty()) {
                results.add(context.resources.getString(R.string.songs))
                results.addAll(songs)
            }

            val artists = ArtistLoader.getArtists(context, searchString)
            if (artists.isNotEmpty()) {
                results.add(context.resources.getString(R.string.artists))
                results.addAll(artists)
            }

            val albums = AlbumLoader.getAlbums(context, searchString)
            if (albums.isNotEmpty()) {
                results.add(context.resources.getString(R.string.albums))
                results.addAll(albums)
            }
            val genres: List<Genre> = GenreLoader.searchGenres(context)
                .filter { genre ->
                    genre.name.toLowerCase(Locale.getDefault())
                        .contains(searchString.toLowerCase(Locale.getDefault()))
                }
            if (genres.isNotEmpty()) {
                results.add(context.resources.getString(R.string.genres))
                results.addAll(genres)
            }
            val playlist = PlaylistLoader.getAllPlaylists(context)
                .filter { playlist ->
                    playlist.name.toLowerCase(Locale.getDefault())
                        .contains(searchString.toLowerCase(Locale.getDefault()))
                }
            if (playlist.isNotEmpty()) {
                results.add(context.getString(R.string.playlists))
                results.addAll(playlist)
            }
        }
        return results
    }
}
